standard/patterns.md

```diff
@@ -0,0 +1,172 @@
+I would suggest placing this new chapter between the existing 12 (Statements) and 13 (namespaces)
+
+# §patterns-new-clause Patterns and pattern matching
+
+## §patterns-new-clause-general General
+
+A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested in the context of a switch expression or a *relational_expression* that is on the left-hand side of an `is` operator. We call this a ***pattern input value***.
+
+## §patterns-new-clause-forms Pattern Forms
+
+A pattern may have one of the following forms:
+
+```ANTLR
+pattern
+    : declaration_pattern
+    | constant_pattern
+    | var_pattern
+    ;
+```
+
+A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.
+
+Each pattern form defines the set of types for input values that that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
+
+Each pattern form defines the set of values for which the pattern *matches* the value.
+
+### §declaration-pattern-new-clause Declaration pattern
+
+A *declaration_pattern* is used to test that a value has a given type and if the test succeeds, to cast that value to that type.
+
+```ANTLR
+declaration_pattern
+    : type simple_designation
+    ;
+simple_designation
+    : single_variable_designation
+    ;
+single_variable_designation
+    : identifier
+    ;
+```
+
+The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.
```

Comment:
This should include boxing and unboxing conversions, right?

Reply:
Yes.  We should unify with the description of the is-type operator.

Reply:
This section should refer to the specification of the is-type operator for the meaning of when it is true or false.  This is probably better done sooner rather than later.

Reply:
Done in #934